rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,217 @@
+# SlateDB Range Scans
+
+<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->
+
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Public API](#public-api)
+- [Implementation](#implementation)
+   * [Checkpoint References](#checkpoint-references)
+   * [Memory Usage](#memory-usage)
+- [Rejected Alternatives & Follow-Ups](#rejected-alternatives-follow-ups)
+- [Updates](#updates)
+
+<!-- TOC end -->
+
+Status: Discussion
+
+Authors:
+
+* [Jason Gustafson](https://github.com/hachikuji)
+
+References:
+
+* https://github.com/slatedb/slatedb/issues/8
+* https://github.com/facebook/rocksdb/wiki/Iterator
+* https://github.com/facebook/rocksdb/wiki/Iterator-Implementation
+* https://github.com/slatedb/slatedb/pull/260
+* https://github.com/slatedb/slatedb/pull/278
+
+## Motivation
+
+
+The ability to scan the keys in a range is an essential feature for building applications.
+For example, this can be used to implement a collections of keys using a common prefix. 
+Currently, SlateDB only supports fetching individual values for known keys.
+
+## Goals
+
+The goal of this work is to provide a flexible capability to query a range of keys in the
+database. It should be possible to specify unbounded ranges to allow querying all keys,
+or bounded ranges for finer precision and efficiency.
+
+Range queries have the potential to be expensive from a resource perspective because they
+hold references to existing tables, which prevents garbage collection. Our goal here is
+to provide a simple initial approach to state management which optimizes for common cases,
+while leaving room for further optimization in the future.
+
+## Public API
+
+We introduce a new `Db::scan` API which accept a `DbRange` defining the range of the scan.
+The result will be a `DbIterator` object which iterates the current records within the range.
+Users may either provide an explicit starting range in the call to `scan`, or they
+may `seek` using the returned iterator.
+
+```rust
+impl Db {
+    /// Scan a range of keys.
+    /// 
+    /// returns a `DbIterator`
+    pub async fn scan(
+        &self,
+        range: std::ops::Range<&[u8]>
+    ) -> Result<DbIterator, SlateDbError>;
+}
+
+impl DbIterator {
+    /// Get the next record in the scan.
+    /// 
+    /// returns Ok(None) when the scan is complete
+    /// returns Err(InvalidatedIterator) if the iterator has been invalidated
```

Comment:
Is the plan to handle all kinds of errors by invalidating the iterator? E.g. suppose the permissions to the bucket change and gets start failing with access denied errors - is that also handled by invalidating the iterator? In that case, would there be some way to get to the underlying error? If not, then what error conditions result in the iterator being invalidated?

Reply:
I guess what I had in mind is reserving that error for cases when the system proactively closes the iterator. For example, if the database is closed or if the checkpoint expires.

Reply:
Got it, so for other errors we'd just return that error directly (which makes sense to me)?